dd-sdk-android/src/main/kotlin/com/datadog/android/sessionreplay/internal/net/BatchesToSegmentsMapper.kt

```diff
@@ -0,0 +1,103 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.android.sessionreplay.internal.net
+
+import com.datadog.android.sessionreplay.model.MobileSegment
+import com.datadog.android.sessionreplay.processor.EnrichedRecord
+import com.datadog.android.sessionreplay.utils.SessionReplayRumContext
+import com.google.gson.JsonArray
+import com.google.gson.JsonObject
+import com.google.gson.JsonParseException
+import com.google.gson.JsonParser
+
+internal class BatchesToSegmentsMapper {
+
+    fun map(batchData: List<ByteArray>): List<Pair<MobileSegment, JsonObject>> {
+        return groupBatchDataIntoSegments(batchData)
+    }
+
+    // region Internal
+
+    private fun groupBatchDataIntoSegments(batchData: List<ByteArray>):
+        List<Pair<MobileSegment, JsonObject>> {
+        return groupBatchDataByRumContext(batchData)
+            .filter { !it.value.isEmpty }
+            .map { entry ->
+                val records = entry.value.map { it.asJsonObject }
+
+                // we are filtering out empty records so we are safe to call first/last functions
+                @Suppress("UnsafeThirdPartyFunctionCall")
+                val startTimestamp = records.first().getAsJsonPrimitive(TIMESTAMP_KEY).asLong
+
+                @Suppress("UnsafeThirdPartyFunctionCall")
+                val stopTimestamp = records.last().getAsJsonPrimitive(TIMESTAMP_KEY).asLong
+                val hasFullSnapshotRecord = hasFullSnapshotRecord(records)
+                val segment = MobileSegment(
+                    MobileSegment.Application(entry.key.applicationId),
+                    MobileSegment.Session(entry.key.sessionId),
+                    MobileSegment.View(entry.key.viewId),
+                    startTimestamp,
+                    stopTimestamp,
+                    records.size.toLong(),
+                    // TODO: RUMM-2518 Find a way or alternative to provide a reliable indexInView
+                    null,
+                    hasFullSnapshotRecord,
+                    MobileSegment.Source.ANDROID,
+                    emptyList()
+                )
+                val segmentAsJsonObject = segment.toJson().asJsonObject
+                segmentAsJsonObject.add(RECORDS_KEY, entry.value)
```

Comment:
nope because that's the catch, the `Segment` accepts a list of `MobileSegment.MobileRecord` and I need to deserialize them back and then to serialize them again together with the `Segment`. Deserializing them back will be a mess as we do not have proper functions for this in the generated classes and we will end up with collisions. Also it will be extra computation for nothing.